torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -932,3 +943,187 @@ def save_checkpoint(
+        should only be used for the intermediate checkpoints. Final checkpoint has to be a synchronous
+        one as the finetuning job can not terminate until the checkpoint gets persisted.
+
+        Args:
+            state_dict (Dict[str, Any]): Checkpoint state dict to be written out to file
+            epoch (int): Epoch number. Used to create the checkpoint file name
+            save_async (bool): If True, save the checkpoint asynchronously
+        """
+
+        log_rank_zero(
+            logger,
+            msg=f"DistributedCheckpointer is saving a checkpoint for the epoch {epoch}",
+        )
+
+        checkpoint_path = Path.joinpath(
+            self._output_dir, f"{self._checkpoint_dir_prefix}_{epoch}"
+        )
+
+        if self._checkpoint_future and not self._checkpoint_future.done():
+            # Previous checkpoint needs to finish before saving the next one.
+            wait_start = time.perf_counter()
+
+            logger.info(
+                f"Rank {self._rank}: previous checkpoint has not finished. Checkpointing frequency is too high. Waiting...",
+            )
+
+            self._checkpoint_future.result()
+
+            logger.info(
+                f"Rank {self._rank}: waited {time.perf_counter() - wait_start:.2f} seconds for previous checkpoint to finish",
+            )
+            self._checkpoint_future = None
+
+        cp_start = time.perf_counter()
+
+        if save_async:
+
+            def callback(
+                f: Future,
+            ) -> None:
+                if f.exception() is None:
+                    logger.info(
+                        f"Rank {self._rank}: Checkpoint is saved asynchronously to {checkpoint_path} successfully.",
+                    )
+                else:
+                    logger.error(
+                        f"Rank {self._rank}: Checkpoint failed to save asynchronously to {checkpoint_path} "
+                        f"with the exception {f.exception()}"
+                    )
+
+            self._checkpoint_future = async_save(
+                state_dict=state_dict,
+                storage_writer=FileSystemWriter(
+                    checkpoint_path,
+                    thread_count=16,
+                    single_file_per_rank=False,
+                    sync_files=False,
+                ),
+                process_group=self._process_group,
+            )
+
+            logger.info(
+                f"Rank {self._rank}: Trainer was blocked for {time.perf_counter() - cp_start:.2f} seconds "
+                "for checkpointing to finish...",
+            )
+
+            self._checkpoint_future.add_done_callback(callback)
+        else:
+            log_rank_zero(
+                logger,
+                msg=f"Saving model checkpoint synchronously to {checkpoint_path}.",
+            )
+
+            save(
+                state_dict=state_dict,
+                storage_writer=FileSystemWriter(
+                    checkpoint_path,
```

Comment:
can we make this API easy to save to manifold internally, so we can avoid rewriting this code? Or thats not how it works?

Reply:
Yes. Lets do that in a follow up PR. Its as simple as plumbing in in FSSpec with Manifold based storage components which already exists. We will need to set up a way to plug in internal components in TorchTune, like we do in TorchTNT etc.